src/peerbook/libp2p_peer.erl

```diff
@@ -93,22 +106,136 @@ metadata_get(Peer=#libp2p_signed_peer_pb{}, Key, Default) ->
+    case lists:keyfind(AssocType, 1, Assocs) of
+        false -> [];
+        {_, #libp2p_association_list_pb{associations=AssocEntries}} -> AssocEntries
+    end.
+
+%% @doc Adds or replaces a given assocation for the given type to the
+%% peer. The returned peer is signed with the provided signing
+%% function.
+-spec associations_put(peer(), AssocType::string(), association(), PeerSigFun::libp2p_crypto:sig_fun())
+                      -> peer().
+associations_put(Peer=#libp2p_signed_peer_pb{}, AssocType, Assoc, PeerSigFun) ->
+    %% Get current associations for type
+    CurrentAssocs = associations_get(Peer, AssocType),
+    %% Store the new association, replacing an existing one if there.
+    UpdatedAssocs = lists:keystore(association_address(Assoc),
+                                   #libp2p_association_pb.address,
+                                   CurrentAssocs,
+                                   Assoc),
+    %% and set the associations for the same type
+    associations_set(Peer, AssocType, UpdatedAssocs, PeerSigFun).
+
+
+%% @doc Checks whether a given peer has an association stored with the
+%% given assocation type and address
+-spec is_association(peer(), AssocType::string(), AssocAddress::libp2p_crypto:address()) -> boolean().
+is_association(Peer=#libp2p_signed_peer_pb{}, AssocType, AssocAddr) ->
+    Assocs = associations_get(Peer, AssocType),
+    case lists:keyfind(AssocAddr, #libp2p_association_pb.address, Assocs) of
+        false -> false;
+        _ -> true
+    end.
+
+%% @doc Make an association for a given peer address. The returned
+%% association contains the given assocation address and the given
+%% peer address signed with the passed in association provided
+%% signature function.
+-spec mk_association(AssocAddr::libp2p_crypto:address(),
+                     PeerAddr::libp2p_crypto:address(),
+                     AssocSigFun::libp2p_crypto:sig_fun()) -> association().
+mk_association(AssocAddr, PeerAddr, AssocSigFun) ->
+    #libp2p_association_pb{ address=AssocAddr,
+                            signature=AssocSigFun(PeerAddr)
+                          }.
+
+%% @doc Gets the address for the given association
+-spec association_address(association()) -> libp2p_crypto:address().
+association_address(#libp2p_association_pb{address=Address}) ->
+    Address.
+
+%% @doc Gets the signature for the given association
+-spec association_signature(association()) -> binary().
+association_signature(#libp2p_association_pb{signature=Signature}) ->
+    Signature.
+
+%% @doc Returns true if the association can be verified against the
+%% given peer address. This has to be the same peer address that was
+%% used to construct the signature in the assocation and should be the
+%% address of the peer record containing the given association.
+-spec association_verify(association(), PeerAddr::libp2p_crypto:address()) -> true.
+association_verify(Assoc=#libp2p_association_pb{}, Address) ->
```

Comment:
Is the user expected to check this, or are peerbook entries checked for this on receipt over gossip?

Reply:
Maybe this doesn't need to be exported?

Reply:
peerbook entries are verified on:

1. peerbook:put which happens during data gossip
2. data decode,so when a peerbook entry is decoded from binary

Reply:
The main reason I exported it was to be symmetric with the higher level verify and to be able to unit test just the association creation logic. Does that sound reasonable?

Reply:
yeah, that makes sense. just checking a user shouldn't have to call it